File: pkg/audit/audit.go
// Package audit writes security-sensitive actions to the audit_logs table.
// Currently used only by pkg/auth, for exactly three actions: user.login
// (success and failure, with a reason), user.refresh_token_reuse (a reused
// refresh token JTI — a signal of a possibly-stolen token), and
// user.logout. Nothing else in the codebase calls into this package yet —
// device credential rotation, alert/incident modifications, and admin
// actions are not audited today, and there is no dead-letter-queue admin
// functionality to audit in the first place (see "Dead-letter queue" in
// README.md). Extending coverage to those would mean calling
// Logger.Log from the relevant handlers; this package's shape doesn't
// need to change to support that, there's just no caller yet.
package audit

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/jackc/pgx/v5/pgxpool"
)

// Entry mirrors the audit_logs schema. OrganizationID/UserID/ResourceID/
// IPAddress/RequestID are pointers because several audit-worthy actions
// (a failed login with an unknown email, a system-initiated action) have no
// value for one or more of these.
type Entry struct {
	OrganizationID *string
	UserID         *string
	Action         string
	ResourceType   string
	ResourceID     *string
	IPAddress      *string
	RequestID      *string
	Result         string // SUCCESS | FAILURE
	Metadata       map[string]any
}

const (
	ResultSuccess = "SUCCESS"
	ResultFailure = "FAILURE"
)

type Logger struct {
	pool *pgxpool.Pool
}

func NewLogger(pool *pgxpool.Pool) *Logger {
	return &Logger{pool: pool}
}

func (l *Logger) Log(ctx context.Context, e Entry) error {
	metadata := e.Metadata
	if metadata == nil {
		metadata = map[string]any{}
	}
	metadataJSON, err := json.Marshal(metadata)
	if err != nil {
		return fmt.Errorf("marshal audit metadata: %w", err)
	}

	_, err = l.pool.Exec(ctx,
		`INSERT INTO audit_logs (organization_id, user_id, action, resource_type, resource_id, ip_address, request_id, result, metadata)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)`,
		e.OrganizationID, e.UserID, e.Action, e.ResourceType, e.ResourceID, e.IPAddress, e.RequestID, e.Result, metadataJSON,
	)
	if err != nil {
		return fmt.Errorf("insert audit log entry (action=%s): %w", e.Action, err)
	}
	return nil
}

File: shared/audit.py
"""Security-sensitive action logging, the Python port of pkg/audit.

Currently used only by shared.auth, for exactly three actions: user.login
(success and failure, with a reason), user.refresh_token_reuse (a reused
refresh token JTI -- a signal of a possibly-stolen token), and user.logout.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from psycopg_pool import ConnectionPool

RESULT_SUCCESS = "SUCCESS"
RESULT_FAILURE = "FAILURE"


@dataclass
class Entry:
    """Mirrors the audit_logs schema. organization_id/user_id/resource_id/
    ip_address/request_id may be None because several audit-worthy actions
    (a failed login with an unknown email, a system-initiated action) have
    no value for one or more of these."""

    action: str
    resource_type: str
    result: str
    organization_id: str | None = None
    user_id: str | None = None
    resource_id: str | None = None
    ip_address: str | None = None
    request_id: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


class Logger:
    def __init__(self, pool: ConnectionPool):
        self._pool = pool

    def log(self, e: Entry) -> None:
        metadata_json = json.dumps(e.metadata or {})
        with self._pool.connection() as conn:
            conn.execute(
                """
                INSERT INTO audit_logs (organization_id, user_id, action, resource_type, resource_id, ip_address, request_id, result, metadata)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
                """,
                (e.organization_id, e.user_id, e.action, e.resource_type, e.resource_id, e.ip_address, e.request_id, e.result, metadata_json),
            )
